server/src/main/java/org/elasticsearch/common/geo/LinearRingEdgeTreeReader.java

```diff
@@ -0,0 +1,148 @@
+        int rightOffset;
+
+        Edge(StreamInput input, int x1, int y1, int x2, int y2, int minY, int maxY, int rightOffset) {
+            this.input = input;
+            this.x1 = x1;
+            this.y1 = y1;
+            this.x2 = x2;
+            this.y2 = y2;
+            this.minY = minY;
+            this.maxY = maxY;
+            this.rightOffset = rightOffset;
+        }
+
+        Edge readLeft() throws IOException {
+            return readEdge(input);
+        }
+
+        Edge readRight() throws IOException {
+            input.skip(rightOffset);
+            return readEdge(input);
+        }
+
+        /** Returns true if the box crosses any edge in this edge subtree */
+        private boolean insideOrCrosses(int minX, int minY, int maxX, int maxY) throws IOException {
+            // we just have to cross one edge to answer the question, so we descend the tree and return when we do.
+            if (this.maxY >= minY) {
+                // is bbox-query contained within linearRing
+                // cast infinite ray to the right from bottom-left and top-right of bbox-query to see if it intersects edge
+                if (lineRelateLine(x1, y1, x2, y2, minX, minY, Integer.MAX_VALUE, minY) != PointValues.Relation.CELL_OUTSIDE_QUERY ||
+                    lineRelateLine(x1, y1, x2, y2, maxX, maxY, Integer.MAX_VALUE, maxY) != PointValues.Relation.CELL_OUTSIDE_QUERY) {
+                    return true;
+                }
+
+                // does rectangle's edges intersect or reside inside polygon's edge
+                if (lineRelateLine(x1, y1, x2, y2, minX, minY, maxX, minY) != PointValues.Relation.CELL_OUTSIDE_QUERY ||
+                    lineRelateLine(x1, y1, x2, y2, maxX, minY, maxX, maxY) != PointValues.Relation.CELL_OUTSIDE_QUERY ||
+                    lineRelateLine(x1, y1, x2, y2, maxX, maxY, minX, maxY) != PointValues.Relation.CELL_OUTSIDE_QUERY ||
+                    lineRelateLine(x1, y1, x2, y2, minX, maxY, minX, minY) != PointValues.Relation.CELL_OUTSIDE_QUERY) {
+                    return true;
+                }
+
+                if (rightOffset > 1) { /* has left node */
+                    if (readLeft().insideOrCrosses(minX, minY, maxX, maxY)) {
+                        return true;
+                    }
+                }
+
+                if (rightOffset > 0 && maxY >= this.minY) { /* no right node if rightOffset == -1 */
+                    if (readRight().insideOrCrosses(minX, minY, maxX, maxY)) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private static Edge readEdge(StreamInput input) throws IOException {
+            int minY = input.readInt();
+            int maxY = input.readInt();
+            int x1 = input.readInt();
+            int y1 = input.readInt();
+            int x2 = input.readInt();
+            int y2 = input.readInt();
+            int rightOffset = input.readInt();
```

Comment:
feels a bit wrong that rightOffset is a property of the edge, it'd be more natural for me to remove it from the edges and let call sites deal with it, e.g.

```
Edge left = readEdge(in);
int rightOffset = in.readInt();
in.skip(rightOffset);
Edge right = readEdge(in);
```

Reply:
agreed. I will pull this info up to the reader!